torch/lib/c10d/logger.cpp

```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
+    // It is possible users did not call backward or run codes in
+    // no-sync mode, in this case, some cudaEvents like "backward_compute_end"
+    // or "backward_comm_start" or "backward_comm_end" will not be recorded.
+    // cudaEvent is created when it is first time to be recorded.
+    // If it is never recorded/created, skip synchronize and calculation.
+    // Otherwise it will throw cuda errors.
+    if (!reducer_->gpu_timer_.forward_start.isCreated() ||
+        !reducer_->gpu_timer_.backward_compute_start.isCreated() ||
+        !reducer_->gpu_timer_.backward_compute_end.isCreated() ||
+        !reducer_->gpu_timer_.backward_comm_start.isCreated() ||
+        !reducer_->gpu_timer_.backward_comm_end.isCreated()) {
+      return;
+    }
+
+    // set_runtime_stats is called at the beginning of forward call,
+    // when it is cheap to synchronize the cuda events of previous iteration,
+    // as mostly all cuda operations are finished in previous iteration.
+    reducer_->gpu_timer_.forward_start.synchronize();
+    reducer_->gpu_timer_.backward_compute_start.synchronize();
+    reducer_->gpu_timer_.backward_compute_end.synchronize();
+    reducer_->gpu_timer_.backward_comm_start.synchronize();
+    reducer_->gpu_timer_.backward_comm_end.synchronize();
+    calculate_avg_gpu_time(
+        ddp_logging_data_->avg_forward_compute_time,
+        reducer_->gpu_timer_.forward_start,
+        reducer_->gpu_timer_.backward_compute_start);
+    calculate_avg_gpu_time(
+        ddp_logging_data_->avg_backward_compute_time,
+        reducer_->gpu_timer_.backward_compute_start,
+        reducer_->gpu_timer_.backward_compute_end);
+    calculate_avg_gpu_time(
+        ddp_logging_data_->avg_backward_comm_time,
+        reducer_->gpu_timer_.backward_comm_start,
+        reducer_->gpu_timer_.backward_comm_end);
+    calculate_avg_gpu_time(
+        ddp_logging_data_->avg_backward_compute_comm_overlap_time,
+        reducer_->gpu_timer_.backward_comm_start,
+        reducer_->gpu_timer_.backward_compute_end);
+#endif
+  } else {
+    calculate_avg_cpu_time(
+        ddp_logging_data_->avg_forward_compute_time,
+        reducer_->cpu_timer_.forward_start_time,
+        reducer_->cpu_timer_.backward_compute_start_time);
+
+    calculate_avg_cpu_time(
+        ddp_logging_data_->avg_backward_compute_time,
+        reducer_->cpu_timer_.backward_compute_start_time,
+        reducer_->cpu_timer_.backward_compute_end_time);
+
+    calculate_avg_cpu_time(
+        ddp_logging_data_->avg_backward_comm_time,
+        reducer_->cpu_timer_.backward_comm_start_time,
+        reducer_->cpu_timer_.backward_comm_end_time);
+
+    calculate_avg_cpu_time(
+        ddp_logging_data_->avg_backward_compute_comm_overlap_time,
+        reducer_->cpu_timer_.backward_comm_start_time,
+        reducer_->cpu_timer_.backward_compute_end_time);
+  }
+}
+
 c10::DDPLoggingData Logger::get_ddp_logging_data() {
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
```

Comment:
Why does getting ddp_logging_data require reducer's mutex?

Reply:
to sync up with threads that are mutating ddp_logging_data